File: Android/melody-core/src/main/java/com/melody/core/schema/ScreenDefinition.kt
package com.melody.core.schema

import kotlinx.serialization.Serializable

@Serializable
data class SearchConfig(
    val stateKey: String,
    val prompt: String? = null,
    val onSubmit: String? = null,
    val placement: String? = null,
    val minimized: Boolean? = null
)

@Serializable
data class ContentInset(
    val top: Double? = null,
    val bottom: Double? = null,
    val leading: Double? = null,
    val trailing: Double? = null,
    val vertical: Double? = null,
    val horizontal: Double? = null
) {
    val resolvedTop: Double get() = top ?: vertical ?: 0.0
    val resolvedBottom: Double get() = bottom ?: vertical ?: 0.0
    val resolvedLeading: Double get() = leading ?: horizontal ?: 0.0
    val resolvedTrailing: Double get() = trailing ?: horizontal ?: 0.0
}

@Serializable
data class ScreenDefinition(
    val id: String,
    val path: String,
    val title: String? = null,
    val titleDisplayMode: String? = null,
    val state: Map<String, StateValue>? = null,
    val onMount: String? = null,
    val body: List<ComponentDefinition>? = null,
    val tabs: List<TabDefinition>? = null,
    val tabStyle: String? = null,
    val toolbar: List<ComponentDefinition>? = null,
    val titleMenu: List<ComponentDefinition>? = null,
    val titleMenuBuilder: String? = null,
    val search: SearchConfig? = null,
    val scrollEnabled: Boolean? = true,
    val wrapper: String? = null,
    val formStyle: String? = null,
    val contentInset: ContentInset? = null,
    val onRefresh: String? = null
)

File: Sources/Core/Schema/Screen/ScreenDefinition.swift
import Foundation


/// A single screen in the app
public struct ScreenDefinition: Codable, Sendable {
    public var id: String
    public var path: String
    public var title: String?
    public var titleDisplayMode: String?
    public var state: [String: StateValue]?
    public var onMount: String?
    public var body: [ComponentDefinition]?
    public var tabs: [TabDefinition]?
    public var tabStyle: String?
    public var toolbar: [ComponentDefinition]?
    public var titleMenu: [ComponentDefinition]?
    public var titleMenuBuilder: String?
    public var search: SearchConfig?
    public var scrollEnabled: Bool?
    public var wrapper: String?
    public var formStyle: String?
    public var contentInset: ContentInset?
    public var onRefresh: String?
    public var showsLoadingIndicator: Bool?

    public init(
        id: String,
        path: String,
        title: String? = nil,
        titleDisplayMode: String? = nil,
        state: [String: StateValue]? = nil,
        onMount: String? = nil,
        body: [ComponentDefinition]? = nil,
        tabs: [TabDefinition]? = nil,
        tabStyle: String? = nil,
        toolbar: [ComponentDefinition]? = nil,
        titleMenu: [ComponentDefinition]? = nil,
        titleMenuBuilder: String? = nil,
        search: SearchConfig? = nil,
        scrollEnabled: Bool? = nil,
        wrapper: String? = nil,
        formStyle: String? = nil,
        contentInset: ContentInset? = nil,
        onRefresh: String? = nil,
        showsLoadingIndicator: Bool? = nil
    ) {
        self.id = id
        self.path = path
        self.title = title
        self.titleDisplayMode = titleDisplayMode
        self.state = state
        self.onMount = onMount
        self.body = body
        self.tabs = tabs
        self.tabStyle = tabStyle
        self.toolbar = toolbar
        self.titleMenu = titleMenu
        self.titleMenuBuilder = titleMenuBuilder
        self.search = search
        self.scrollEnabled = scrollEnabled ?? true
        self.wrapper = wrapper
        self.formStyle = formStyle
        self.contentInset = contentInset
        self.onRefresh = onRefresh
        self.showsLoadingIndicator = showsLoadingIndicator ?? true
    }
}
